core/src/main/java/org/elasticsearch/index/percolator/PercolatorQueryCache.java

```diff
@@ -0,0 +1,278 @@
+        QueriesLeaf percolatorQueries = cache.get(ctx.reader().getCoreCacheKey());
+        if (percolatorQueries == null) {
+            throw new IllegalStateException("queries not loaded, queries should be have been preloaded during index warming...");
+        }
+        return percolatorQueries;
+    }
+
+    public IndexWarmer.Listener createListener(ThreadPool threadPool) {
+        return new IndexWarmer.Listener() {
+
+            final Executor executor = threadPool.executor(ThreadPool.Names.WARMER);
+
+            @Override
+            public IndexWarmer.TerminationHandle warmNewReaders(IndexShard indexShard, Engine.Searcher searcher) {
+                final CountDownLatch latch = new CountDownLatch(searcher.reader().leaves().size());
+                for (final LeafReaderContext ctx : searcher.reader().leaves()) {
+                    executor.execute(() -> {
+                        try {
+                            final long start = System.nanoTime();
+                            QueriesLeaf queries = loadQueries(ctx, indexShard.indexSettings().getIndexVersionCreated());
+                            cache.put(ctx.reader().getCoreCacheKey(), queries);
+                            if (indexShard.warmerService().logger().isTraceEnabled()) {
+                                indexShard.warmerService().logger().trace(
+                                        "loading percolator queries took [{}]",
+                                        TimeValue.timeValueNanos(System.nanoTime() - start)
+                                );
+                            }
+                        } catch (Throwable t) {
+                            indexShard.warmerService().logger().warn("failed to load percolator queries", t);
+                        } finally {
+                            latch.countDown();
+                        }
+                    });
+                }
+                return () -> latch.await();
+            }
+
+            @Override
+            public IndexWarmer.TerminationHandle warmTopReader(IndexShard indexShard, Engine.Searcher searcher) {
+                return IndexWarmer.TerminationHandle.NO_WAIT;
+            }
+        };
+    }
+
+    QueriesLeaf loadQueries(LeafReaderContext context, Version indexVersionCreated) throws IOException {
+        LeafReader leafReader = context.reader();
+        ShardId shardId = ShardUtils.extractShardId(leafReader);
+        if (shardId == null) {
+            throw new IllegalStateException("can't resolve shard id");
+        }
+        if (indexSettings.getIndex().equals(shardId.getIndex()) == false) {
+            // percolator cache insanity
+            String message = "Trying to load queries for index " + shardId.getIndex() + " with cache of index " + indexSettings.getIndex();
+            throw new IllegalStateException(message);
+        }
+
+        int numQueries = 0;
+        Query[] queries = new Query[leafReader.maxDoc()];
```

Comment:
actually I'm wondering if we should use an IntObjectHashMap: given that the percolator works on a type, I'm afraid users have a lot of data in other types too so creating a Query[maxDoc] could cause an OOME on some of them. By using an IntObjectHashMap we will be more on the safe side: it will une more memory if all docs have a query but much less in the sparse case.